File: js/protocol.js
// CoolLED1248 / CoolLEDX BLE wire protocol.
//
// Reimplemented from the community reverse-engineering of the protocol used
// by "CoolLEDX" flexible LED matrix signs (the ones paired via the
// CoolLED1248 phone app). Reference: CrimsonClyde's LED FaceShields project
// and the UpDryTwist/coolledx-driver Python port.

export const SERVICE_UUID = '0000fff0-0000-1000-8000-00805f9b34fb';
export const CHARACTERISTIC_UUID = '0000fff1-0000-1000-8000-00805f9b34fb';

export const CMD = {
  MUSIC: 0x01,
  TEXT: 0x02,
  IMAGE: 0x03,
  ANIMATION: 0x04,
  ICON: 0x05,
  MODE: 0x06,
  SPEED: 0x07,
  BRIGHTNESS: 0x08,
  SWITCH: 0x09,
  TRANSFER: 0x0a,
  INVERT_DISPLAY: 0x0c,
  CLEAR_MAYBE: 0x0d,
  SHOW_ICON: 0x11,
  POWER_DOWN: 0x12,
  POWER_ON: 0x13,
  INVERT_OR_SOMETHING: 0x15,
  INITIALIZE: 0x23,
};

export const MODE = {
  STATIC: 0x01,
  LEFT: 0x02,
  RIGHT: 0x03,
  UP: 0x04,
  DOWN: 0x05,
  SNOWFLAKE: 0x06,
  PICTURE: 0x07,
  LASER: 0x08,
};

const CHUNK_SIZE = 128;

function u16be(n) {
  return [(n >> 8) & 0xff, n & 0xff];
}

// Framing bytes (0x01, 0x02, 0x03) that appear inside the payload must be
// escaped so the receiver can tell them apart from the start/end markers.
// 0x02 -> 0x02 0x06, 0x01 -> 0x02 0x05, 0x03 -> 0x02 0x07 (order matters:
// 0x02 must be escaped first so newly-introduced 0x02 bytes aren't re-escaped).
function escapeBytes(bytes) {
  const out = [];
  for (const b of bytes) {
    if (b === 0x02) out.push(0x02, 0x06);
    else if (b === 0x01) out.push(0x02, 0x05);
    else if (b === 0x03) out.push(0x02, 0x07);
    else out.push(b);
  }
  return out;
}

// Wrap a payload as 0x01 [escaped 2-byte-length + payload] 0x03.
function createCommand(payloadBytes) {
  const withLength = [...u16be(payloadBytes.length), ...payloadBytes];
  const escaped = escapeBytes(withLength);
  return new Uint8Array([0x01, ...escaped, 0x03]);
}

function xorChecksum(bytes) {
  let c = 0;
  for (const b of bytes) c ^= b;
  return c;
}

function splitBytes(data, chunkSize) {
  const chunks = [];
  for (let i = 0; i < data.length; i += chunkSize) {
    chunks.push(data.slice(i, i + chunkSize));
  }
  return chunks.length ? chunks : [data];
}

// Large payloads (text/image/animation pixel data) get split into <=128
// byte chunks, each with a small header (unknown 0x00, total length,
// chunk index, chunk size) and an XOR checksum, all prefixed with the
// command byte.
function chopUpData(data, commandByte) {
  const rawChunks = splitBytes(data, CHUNK_SIZE);
  return rawChunks.map((chunk, idx) => {
    const body = [0x00, ...u16be(data.length), ...u16be(idx), chunk.length & 0xff, ...chunk];
    body.push(xorChecksum(body));
    return new Uint8Array([commandByte, ...body]);
  });
}

function simplePacket(cmdByte, ...valueBytes) {
  return [createCommand(new Uint8Array([cmdByte, ...valueBytes]))];
}

export function buildSpeedPackets(speed) {
  return simplePacket(CMD.SPEED, speed & 0xff);
}

export function buildBrightnessPackets(brightness) {
  return simplePacket(CMD.BRIGHTNESS, brightness & 0xff);
}

export function buildModePackets(mode) {
  return simplePacket(CMD.MODE, mode & 0xff);
}

export function buildPowerPackets(on) {
  return simplePacket(CMD.SWITCH, on ? 0x01 : 0x00);
}

export function buildInvertDisplayPackets(inverted) {
  return simplePacket(CMD.INVERT_DISPLAY, inverted ? 0x01 : 0x00);
}

export function buildInitializePackets() {
  return simplePacket(CMD.INITIALIZE, 0x01);
}

export function buildPowerDownPackets() {
  return [createCommand(new Uint8Array([CMD.POWER_DOWN]))];
}

// Build the full multi-chunk wire packets for a rendered text/image payload
// (the "24 zero bytes + text length + char metadata + pixel length + pixel
// bits" structure the sign expects), given already rendered RGB bitfields.
export function buildTextPackets(text, pixelBits) {
  const payload = [];
  payload.push(...new Array(24).fill(0));

  const codepoints = [...text];
  const textLen = codepoints.length;
  let bufferLength = 80;
  if (textLen > 255) {
    payload.push((textLen >> 8) & 0xff, textLen & 0xff);
    bufferLength = 79;
  } else {
    payload.push(textLen & 0xff);
  }
  const meta = new Array(bufferLength).fill(0);
  for (let i = 0; i < Math.min(textLen, bufferLength); i++) meta[i] = 0x30;
  payload.push(...meta);

  payload.push(...u16be(pixelBits.length));
  payload.push(...pixelBits);

  const rawChunks = chopUpData(new Uint8Array(payload), CMD.TEXT);
  return rawChunks.map(createCommand);
}

export function buildImagePackets(pixelBits) {
  const payload = [];
  payload.push(...new Array(24).fill(0));
  payload.push(...u16be(pixelBits.length));
  payload.push(...pixelBits);

  const rawChunks = chopUpData(new Uint8Array(payload), CMD.IMAGE);
  return rawChunks.map(createCommand);
}

File: desktop/led_bridge/protocol.py
"""CoolLED1248 / CoolLEDX BLE wire protocol.

Ported from the browser implementation in js/protocol.js (already
byte-verified against the reference driver's framing/escaping/chunking
logic). Kept deliberately dependency-free so it can be unit tested without
a Bluetooth adapter.
"""

from __future__ import annotations

SERVICE_UUID = "0000fff0-0000-1000-8000-00805f9b34fb"
CHARACTERISTIC_UUID = "0000fff1-0000-1000-8000-00805f9b34fb"

CMD_TEXT = 0x02
CMD_MODE = 0x06

MODE_STATIC = 0x01
MODE_LEFT = 0x02
MODE_RIGHT = 0x03
MODE_UP = 0x04
MODE_DOWN = 0x05
MODE_SNOWFLAKE = 0x06
MODE_PICTURE = 0x07
MODE_LASER = 0x08

_CHUNK_SIZE = 128


def _escape_bytes(data: bytes) -> bytes:
    """Escape framing bytes (0x01/0x02/0x03) inside a payload.

    0x02 -> 0x02 0x06, 0x01 -> 0x02 0x05, 0x03 -> 0x02 0x07. 0x02 must be
    escaped first so newly-introduced 0x02 bytes aren't re-escaped.
    """
    out = bytearray()
    for b in data:
        if b == 0x02:
            out += b"\x02\x06"
        elif b == 0x01:
            out += b"\x02\x05"
        elif b == 0x03:
            out += b"\x02\x07"
        else:
            out.append(b)
    return bytes(out)


def create_command(payload: bytes) -> bytes:
    """Wrap a payload as 0x01 [escaped 2-byte-length + payload] 0x03."""
    with_length = len(payload).to_bytes(2, "big") + payload
    escaped = _escape_bytes(with_length)
    return b"\x01" + escaped + b"\x03"


def _xor_checksum(data: bytes) -> int:
    checksum = 0
    for b in data:
        checksum ^= b
    return checksum


def _split_bytes(data: bytes, chunk_size: int) -> list[bytes]:
    chunks = [data[i : i + chunk_size] for i in range(0, len(data), chunk_size)]
    return chunks or [data]


def _chop_up_data(data: bytes, command_byte: int) -> list[bytes]:
    """Split a large payload into <=128 byte chunks with header + XOR checksum."""
    raw_chunks = _split_bytes(data, _CHUNK_SIZE)
    out = []
    for idx, chunk in enumerate(raw_chunks):
        body = bytearray()
        body.append(0x00)
        body += len(data).to_bytes(2, "big")
        body += idx.to_bytes(2, "big")
        body.append(len(chunk) & 0xFF)
        body += chunk
        body.append(_xor_checksum(body))
        out.append(bytes([command_byte]) + bytes(body))
    return out


def build_mode_packets(mode: int) -> list[bytes]:
    return [create_command(bytes([CMD_MODE, mode & 0xFF]))]


def build_text_packets(text: str, pixel_bits: bytes) -> list[bytes]:
    """Build the wire packets for a text/image payload given rendered pixel bits."""
    payload = bytearray(24)  # unknown 24 zero bytes

    codepoints = list(text)
    text_len = len(codepoints)
    if text_len > 255:
        payload += text_len.to_bytes(2, "big")
        buffer_length = 79
    else:
        payload.append(text_len & 0xFF)
        buffer_length = 80

    meta = bytearray(buffer_length)
    for i in range(min(text_len, buffer_length)):
        meta[i] = 0x30
    payload += meta

    payload += len(pixel_bits).to_bytes(2, "big")
    payload += pixel_bits

    raw_chunks = _chop_up_data(bytes(payload), CMD_TEXT)
    return [create_command(c) for c in raw_chunks]
